### PR TITLE
Added type hints to State, and made it a dataclass

### PR DESCRIPTION
**Pull request by KessemClein**
(trying to make it more newcomer friendly)

### CODE REVIEW

**Comment by ArielHorwitz** [comment]
Looks great to me.

**Comment by imcrazeegamer** [comment]
interesting use of data classes

**Comment by ArielHorwitz** [comment]
I'll just add some thoughts (I don't call the shots on this anyway). Honestly I like the type hinting but I'm not a fan of the use of dataclass in this particular case.

As I understand, dataclasses are designed to save you time from boilerplate code. They may save you from writing some common methods (such as `__eq__` and `__repr__`), however this change only moves attribute definitions from the init to the class namespace (and provides a repr which can be easily written). I believe the State class is intricate (read: specific) enough that it is worth skipping the dataclass and using the useful type hinting. Is my impression valid?

**Comment by KessemClein** [comment]
> 
The` __init__` is the boilerplate we're avoiding here...
I personally don't see dataclasses as a way to avoid boilerplate code (although they can be useful for that) but more of a way to specify a class whose central purpose is to hold a specific type of data. The most important thing a dataclass saves you is the `__init__` method: not only writing it, but it lets you to separate class api from the implementation.

Moving the specification of that data to the class definition as opposed to the implementation of the init function is exactly the point: `positions : List[Hexagon]` is a better way to specify that this class holds a list of `Hexagons` named `positions` than expecting the user to look inside my init method, and seeing `self.positions = positions` and then following the definition of `positions` back to to the arguments. 

Basically, a dataclass gives you a neat specification (or interface) of the class that is decoupled from the implementations, whereas an `__init__` only _specifies the arguments_ used to initialize a state, and the interface is hidden in the implementation (and if we're lucky it's _only_ in the `__init__`).

**Comment by ArielHorwitz** [comment]
This would be my favourite point:
> Basically, a dataclass gives you a neat specification (or interface) of the class that is decoupled from the implementations

**Comment by ArielHorwitz** [comment]
I appreciate all the feedback on how to make the project more newcomer friendly.

The users should have a similar experience with the type hints in #55 . If they would like to study the module I believe the arguments in the init should work for them (though I do intend to add more details in the docstring on each attribute). TBH it looks more readable to me like so, however I am not against the idea in principle. So I will look at an updated PR for making state a dataclass if it comes. And if someone wishes to maintain the module, I would be more than happy to be less involved :)
